src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Refactoring.cs

```diff
@@ -0,0 +1,89 @@
+
+            var severity = GetOptionSeverity(optionValue);
+            switch (severity)
+            {
+                case ReportDiagnostic.Suppress:
+                case ReportDiagnostic.Hidden:
+                    // if the severity is Hidden that's equivalent to 'refactoring only', so we want
+                    // to try to compute the refactoring here.
+                    //
+                    // If the severity is 'suppress', that means the user doesn't want the actual
+                    // analyzer to run here.  However, we can still check to see if we could offer
+                    // the feature here as a refactoring.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, analyzerActive: false).ConfigureAwait(false);
+                    return;
+
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
+                    // User has this option set at a level where we want it checked by the
+                    // DiagnosticAnalyser and not the CodeRefactoringProvider.  However, we still
+                    // want to check if we want to offer the *reverse* refactoring here in this
+                    // single location.
+                    //
+                    // For example, say this is the "use expression body" feature.  If the user says
+                    // they always prefer expression-bodies (with warning level), then we want the
+                    // analyzer to always be checking for that.  However, we still want to offer the
+                    // refactoring to flip their code to use a block body here, just in case that
+                    // was something they wanted to do as a one off (i.e. before adding new
+                    // statements.
+                    //
+                    // TODO(cyrusn): Should we only do this for warn/info?  Argument could be made
+                    // that we shouldn't even offer to refactor in the reverse direction if it will
+                    // just cause an error.  That said, maybe this is just an intermediary step, and
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, analyzerActive: true).ConfigureAwait(false);
+                    return;
+            }
+        }
+
+        private async Task ComputeRefactoringsAsync(
+            CodeRefactoringContext context, TOptionKind option, bool analyzerActive)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            var computationTask = analyzerActive
+                ? ComputeOpposingRefactoringsWhenAnalyzerActiveAsync(document, span, option, cancellationToken)
+                : ComputeRefactoringsWhenAnalyzerInactiveAsync(document, span, cancellationToken);
```

Comment:
I believe it's because you need hte actual option's current value to determine what the 'opposing refactorigns would be'.   Since you're gicing the *opposite* direction of whatever the analyzer is doing based on the current option.

But if the analyzer is completely inactive, you don't care what teh value is for the analyzer.  Becuase you're always offering all available refactorings that you can since the value of the option doesn't affect the refactoring itself.

Reply:
Hard to explain in the abstract, so let's use a real world example.   Say you have "prefer block body" as "true:info".  In this case the analyzer is responsible for all the lightbulbs that offer to convert from expression to block body.  The refactoring then is only responsible for the 'opposite'.  i.e. offering the screwdriver on "convert from block to expression body".

However, say you have this set to "true:none".  Then the analyzer won't ever run since the feature is in 'refactoring mode only'.  In that case, the refactoring will light up on *both* expression and block bodies, and will offer hte screwdriver to convert to the other form.

--

This pattern is one of the reasons this PR is useful.  It codifies and enforces that.  Normally, the refactoring provider should never need to know the option value.  Because the idea of refactorings is that (as long as things aren't completely disabled) they are offered in every direction to be useful.